Core: Support Custom Table/View Operations in RESTCatalog

XJDKC opened the pull request:
Currently, RESTCatalog allows users to replace components such as `RESTClient`, `FileIO`, `AuthManager`, and `MetricsReporter` (with the logic handled in `RESTSessionCatalog`). 
* `RESTClient`: via builder
    * [RESTCatalog.java#L65-L66](https://github.com/apache/iceberg/blob/571b696106f4748d3edee9ee641e214377ddcb16/core/src/main/java/org/apache/iceberg/rest/RESTCatalog.java#L65-L66)
* `FileIO`: via builder (doesn't allow `RESTCatalog` to pass in the builder) or reflection
    * [RESTSessionCatalog.java#L987-L999](https://github.com/apache/iceberg/blob/571b696106f4748d3edee9ee641e214377ddcb16/core/src/main/java/org/apache/iceberg/rest/RESTSessionCatalog.java#L987-L999)
* `AuthManager`: via reflection
    * [RESTSessionCatalog.java#L191](https://github.com/apache/iceberg/blob/571b696106f4748d3edee9ee641e214377ddcb16/core/src/main/java/org/apache/iceberg/rest/RESTSessionCatalog.java#L191)
    * [AuthManagers.java#L87-L117](https://github.com/apache/iceberg/blob/571b696106f4748d3edee9ee641e214377ddcb16/core/src/main/java/org/apache/iceberg/rest/auth/AuthManagers.java#L87-L117)
* `MetricsReporter`: via reflection
    * [RESTSessionCatalog.java#L252](https://github.com/apache/iceberg/blob/571b696106f4748d3edee9ee641e214377ddcb16/core/src/main/java/org/apache/iceberg/rest/RESTSessionCatalog.java#L252)
    * [CatalogUtil.java#L500-L533](https://github.com/apache/iceberg/blob/571b696106f4748d3edee9ee641e214377ddcb16/core/src/main/java/org/apache/iceberg/CatalogUtil.java#L500-L533)

However, one dependent component that remains non-injectable is `RESTTableOperations`.

This PR adds support for injecting custom implementations of table and view operations in `RESTCatalog`, enabling users to extend and customize REST catalog behavior more easily. It doesn't change any functionalities.

This PR also allows user extends `RESTCatalog` and `RESTSessionCatalog` to provide a custom table / view operations by overriding `newTableOps` in `RESTSessionCatalog`

```java
public RESTCatalog(
      SessionCatalog.SessionContext context,
      Function<Map<String, String>, RESTClient> clientBuilder) {
    this.sessionCatalog = newSessionCatalog(clientBuilder);
    // .....
}

protected RESTSessionCatalog newSessionCatalog(
      Function<Map<String, String>, RESTClient> clientBuilder) {
    return new RESTSessionCatalog(clientBuilder, null);
}
```

sfc-gh-mparmar commented on core/src/main/java/org/apache/iceberg/rest/RESTOperationsBuilder.java:
```diff
@@ -0,0 +1,145 @@
+ * RESTSessionCatalog catalog = new RESTSessionCatalog(clientBuilder, ioBuilder, customBuilder);
+ * </pre>
+ */
+public interface RESTOperationsBuilder {
```
Could you add few unit tests by creating a TestRESTOperationsBuilder and injecting it into the RESTCatalog?

XJDKC commented on core/src/main/java/org/apache/iceberg/rest/RESTOperationsBuilder.java:
```diff
@@ -0,0 +1,145 @@
+   * @param endpoints the set of supported REST endpoints
+   * @return a new RESTViewOperations instance
+   */
+  default RESTViewOperations createViewOperations(
```
IIUC, `FileIO` is not required for view operations, because Iceberg views are logical objects that contain only metadata (SQL definitions, schemas, and versions) and do not read or write any physical files.

When a user runs a query against a view, the query engine expands the view's SQL definition, compiles it into a query plan, and resolves the underlying tables. At that point, the engine loads the actual table objects (which include `TableOperations` and `FileIO`) to read the physical data files.

XJDKC commented:
cc: @flyrain @stevenzwu @huaxingao Could you pls take a look when you get a chance? Thanks! 🙏

flyrain commented on core/src/main/java/org/apache/iceberg/rest/RESTCatalog.java:
```diff
@@ -63,13 +65,28 @@ public RESTCatalog() {
+      BiFunction<SessionCatalog.SessionContext, Map<String, String>, FileIO> ioBuilder,
+      RESTOperationsBuilder operationsBuilder) {
+    this(SessionCatalog.SessionContext.createEmpty(), clientBuilder, ioBuilder, operationsBuilder);
+  }
```
Is this method necessary if we go with `this(SessionCatalog.SessionContext.createEmpty(), clientBuilder);` in line 68?

XJDKC replied:
Even though RESTSessionCatalog allows us to pass in the ioBuilder, but the RESTCatalog doesn't, so I add it in the constructor of RESTCatalog as well.

flyrain commented on core/src/main/java/org/apache/iceberg/rest/RESTViewOperations.java:
```diff
@@ -29,14 +29,14 @@
 import org.apache.iceberg.view.ViewOperations;
 
-class RESTViewOperations implements ViewOperations {
+public class RESTViewOperations implements ViewOperations {
```
Do we need this scope change?

XJDKC replied:
Same as above

flyrain reviewed (commented):
Thanks @XJDKC for the change. Left some comments.

XJDKC commented on core/src/main/java/org/apache/iceberg/rest/RESTCatalog.java:
```diff
@@ -63,13 +65,28 @@ public RESTCatalog() {
 
   public RESTCatalog(Function<Map<String, String>, RESTClient> clientBuilder) {
-    this(SessionCatalog.SessionContext.createEmpty(), clientBuilder);
+    this(SessionCatalog.SessionContext.createEmpty(), clientBuilder, null, null);
```
Makes sense, let me revise it.

XJDKC replied:
Done!

XJDKC commented on core/src/main/java/org/apache/iceberg/rest/RESTOperationsBuilder.java:
```diff
@@ -0,0 +1,145 @@
+ * RESTSessionCatalog catalog = new RESTSessionCatalog(clientBuilder, ioBuilder, customBuilder);
+ * </pre>
+ */
+public interface RESTOperationsBuilder {
```
Makes sense, let me revise it!

XJDKC replied:
Done!

XJDKC commented:
> A couple of questions wrt encrypted tables,
> 
> 1. What if the encryption.key-id table property is set (https://github.com/apache/iceberg/blob/main/core/src/main/java/org/apache/iceberg/TableProperties.java#L391) but a custom TO implementation ignores it. Do the users expect a table to be encrypted if the encryption.key-id is set? Should the implementors of custom TOs validate them by running the Iceberg unitests (inc TestTableEncryption)?
> 2. The standard RESTableOperations class, built in Iceberg, uses a safe approach to getting the metadata object (directly from the REST catalog server, never from the metadata.json file that can be kept in untrusted storage). Can custom TO replacements behave differently in this respect?
> 
> If any of these points is a concern, then I believe it can be addressed just by adding a few lines to the RESTOperationsBuilder javadoc API comments. What do you think?

> Should the implementors of custom TOs validate them by running the Iceberg unit tests?

For a custom TableOperations (TO) implementation, the responsibility lies entirely with the implementer. They must ensure proper handling of encryption and any other security-sensitive logic.

That said, even for a custom implementation, I'd expect most of the core logic to remain unchanged and continue to rely on the unmanaged components provided by Iceberg sdk.

We can add some comments or documentation notes to call this out explicitly, so that anyone implementing a custom TableOperations is aware of the encryption keys and understands the need to handle encryption properly (IIRC, in your PR, an additional param will be passed). This should help prevent misuse or accidental security gaps when extending the default implementation.

That said, if someone chooses to extend the default TO, they should take full responsibility for doing so safely. The same applies to the ClientBuilder: users may provide their own HttpClient (for example, to support custom logic (shared connection pool, PrivateLink, proxy, or mTLS, ), and it’s their responsibility to ensure it doesn’t break core functionality.

> The standard RESTableOperations class, built in Iceberg, uses a safe approach to getting the metadata object (directly from the REST catalog server, never from the metadata.json file that can be kept in untrusted storage). Can custom TO replacements behave differently in this respect?

As mentioned earlier, that’s already possible even without this PR. Anyone can build their own library or copy the Iceberg SDK code and modify it as they wish. Iceberg is a specification, and the Apache Iceberg repository serves as a reference implementation, we can't prevent developers from customizing it.

stevenzwu commented on core/src/main/java/org/apache/iceberg/rest/RESTOperationsFactory.java:
```diff
@@ -0,0 +1,145 @@
+  }
+
+  /** Default {@link RESTOperationsFactory} instance. */
+  RESTOperationsFactory DEFAULT = new RESTOperationsFactory() {};
```
nit: we typically define this near the top of the class and make it static

XJDKC commented on core/src/main/java/org/apache/iceberg/rest/RESTOperationsFactory.java:
```diff
@@ -0,0 +1,145 @@
+ * RESTSessionCatalog catalog = new RESTSessionCatalog(clientBuilder, ioBuilder, customFactory);
+ * </pre>
+ */
+public interface RESTOperationsFactory {
```
Hey @amogh-jahagirdar, thanks for reviewing the PR, what you mentioned makes sense to me.

The reason I went with a builder/factory approach is mainly to stay consistent with the existing pattern, since we already have builders for components like `RESTClient` and `FileIO`.

That said, I'm fine with either direction. If we go with your suggested approach, we need a builder / factory for `RESTSessionCatalog` or maybe a method called `newSessionCatalog`, and users need to extend the `RESTSessionCatalog` to provide the custom `TableOperations`.

Actually, I think that’s a cleaner solution and provides more flexibility overall. WDYT?

cc: @stevenzwu @flyrain

XJDKC replied:
Thanks all for the valuable inputs, let me revise my PR based on the inheritance-based approach!

flyrain replied:
I think a cleaner approach would be to keep the member variables private but expose them via protected getter methods.

That way, subclasses of `RESTSessionCatalog` can access the state they need when customizing `newTableOps` or `newViewOps`, but we still preserve encapsulation and avoid committing to the exact field layout long-term.

stevenzwu commented on core/src/main/java/org/apache/iceberg/rest/RESTSessionCatalog.java:
```diff
@@ -979,11 +980,11 @@ private String fullTableName(TableIdentifier ident) {
   }
 
-  private FileIO newFileIO(SessionContext context, Map<String, String> properties) {
+  protected FileIO newFileIO(SessionContext context, Map<String, String> properties) {
```
do we need this scope change for FileIO? Constructor already accepts `ioBuilder` for customization

XJDKC commented:
Hi @stevenzwu @flyrain @amogh-jahagirdar, kindly bumping this PR for visibility.
If everything looks good, could we please go ahead and merge it? Feel free to leave any additional comments or suggestions as well! Thanks!

gaborkaszab commented on core/src/main/java/org/apache/iceberg/rest/RESTSessionCatalog.java:
```diff
@@ -1010,6 +1011,82 @@ private FileIO tableFileIO(
+   * @param supportedEndpoints the set of supported REST endpoints
+   * @return a new RESTTableOperations instance
+   */
+  protected RESTTableOperations newTableOps(
```
I'm just curious here: These are protected, so not part of any API or so. This means that there are no guarantees the signature of these functions don't change and any change here is not considered a breaking change. I don't think revAPI checks would break if I add a couple of extra parameters and change the types for some of the existing ones.
However, if you make your application so that it overrides these functions it could break with every Iceberg version bump. Isn't this an issue at your end?

XJDKC replied:
If someone decides to override this method, they're responsible for adapting to future changes. The Iceberg spec is continuously evolving, even if we define an API today, it doesn't mean it will remain unchanged.
For example, there's another ongoing PR that adds encryption-related properties to `RESTTableOperations`. So it’s quite common for APIs in this area to evolve over time.

But the benefits of just injecting a custom table operations is: if the new changes are not related to this method, we can upgrade smoothly.

gaborkaszab reviewed (commented):
Thank you for the PR and for answering my earlier questions about the use-case, @XJDKC!
I see this is anticipated by a number of community members, so also fine by me. Just scratching the use-case a bit further, you mentioned extra logging and adding extra headers. Wouldn't it make sense to add the logging to the RESTTableOperations code and inject a header supplier or such that could take care of the extra headers. I know this is pretty late to the conversation, sorry, I don't intend to hold this back. Mainly asking for my own information so that I can understand the use-case better.

I added some minor nits, and asked also for my information about the API stability guarantees for engines overriding the protected methods.
Thanks!

nastra commented on core/src/main/java/org/apache/iceberg/rest/RESTTableOperations.java:
```diff
@@ -44,10 +44,10 @@
 import org.apache.iceberg.util.LocationUtil;
 
-class RESTTableOperations implements TableOperations {
+public class RESTTableOperations implements TableOperations {
```
I don't think we should update the visibility here. If somebody wants to extend this class then that is still possible. The subclass just needs to be inside the `org.apache.iceberg.rest` package

XJDKC replied:
Reverted the changes to RESTTableOperations!

nastra commented on core/src/main/java/org/apache/iceberg/rest/RESTSessionCatalog.java:
```diff
@@ -1010,6 +1011,82 @@ private FileIO tableFileIO(
+   * @param path the REST path for the table
+   * @param headers a supplier for additional HTTP headers to include in requests
+   * @param fileIO the FileIO implementation for reading and writing table metadata and data files
+   * @param updateType the type of update being performed (CREATE, REPLACE, or SIMPLE)
```
```suggestion
   * @param updateType the {@link org.apache.iceberg.rest.RESTTableOperations.UpdateType} being performed
```

no need to list all values of the enum

XJDKC commented on core/src/main/java/org/apache/iceberg/rest/RESTSessionCatalog.java:
```diff
@@ -1010,6 +1011,82 @@ private FileIO tableFileIO(
+   * @param supportedEndpoints the set of supported REST endpoints
+   * @return a new RESTTableOperations instance
+   */
+  protected RESTTableOperations newTableOpsForTransaction(
```
Makes sense! I will revise it to just use `newTableOps`.

XJDKC commented on core/src/test/java/org/apache/iceberg/rest/TestRESTCatalog.java:
```diff
@@ -3066,6 +3074,113 @@ public void testCommitStateUnknownNotReconciled() {
+      catalog.createNamespace(ns);
+
+      catalog.createTable(TableIdentifier.of(ns, "table1"), SCHEMA);
+      assertThat(customTableOps).isTrue();
```
Makes sense!

nastra commented on core/src/test/java/org/apache/iceberg/rest/TestRESTCatalog.java:
```diff
@@ -3066,6 +3073,93 @@ public void testCommitStateUnknownNotReconciled() {
+          List<MetadataUpdate> createChanges,
+          TableMetadata current,
+          Set<Endpoint> supportedEndpoints) {
+        customTableOpsCalled.set(true);
```
maybe use two different variables like you had before to distinguish between which one was actually called

gaborkaszab commented on core/src/test/java/org/apache/iceberg/rest/TestRESTCatalog.java:
```diff
@@ -3066,6 +3073,101 @@ public void testCommitStateUnknownNotReconciled() {
 
+  @Test
+  public void testCustomTableOperationsInjection() throws IOException {
+    AtomicBoolean customTableOpsCalled = new AtomicBoolean();
```
Can't we use Mockito.spy to verify that the functions are called instead of introducing booleans for this?

XJDKC replied:
I prefer using boolean flags over `Mockito.verify` for this test because:
1. More realistic: The spy approach requires exposing and wrapping the session catalog instance in `newSessionCatalog()`, which isn't how users would actually extend these classes.
2. Test the right thing: The boolean verifies that our custom implementation runs. `newTableOps()` will be called either way (default or custom), so `verify()` doesn't actually prove customization worked.
3. Simpler and straight forward: set a flag in the override to prove it executed. No mocking scaffolding needed. 

gaborkaszab commented on core/src/test/java/org/apache/iceberg/rest/TestRESTCatalog.java:
```diff
@@ -3066,6 +3073,101 @@ public void testCommitStateUnknownNotReconciled() {
+          TableMetadata current,
+          Set<Endpoint> supportedEndpoints) {
+        customTableOpsCalled.set(true);
+        return super.newTableOps(restClient, path, headers, fileIO, current, supportedEndpoints);
```
Somehow my comment got lost amongst other comments so let me repeat it here separately:

I checked the test for this PR and apparently we test how to put a wrapper around the ops creation by overriding RESTSessionCatalog and RESTCatalog functionality, but we don't test the end-to-end use case. I think injecting a custom table/view ops would be nice to see how this would work. Like injecting an ops that adds extra header to requests and also verify that these headers are present, since that was the original motivation of this PR.

gaborkaszab reviewed (commented):
Thank you for your answers, @XJDKC !

XJDKC commented:
Thanks @flyrain @stevenzwu @nastra @amogh-jahagirdar @gaborkaszab for reviewing this PR and sharing your feedback.

I've addressed all the comments and also added a more comprehensive test for the custom table/view operations. If there are no other comments, maybe we can go ahead and merge this PR? Thanks!
